pkg/ha/filter.go

```diff
@@ -49,42 +51,125 @@ func (h *Filter) Process(_ *http.Request, wr *prompb.WriteRequest) error {
+	switch {
+	case !hasBackfill && allowInsert:
+		filterOutSampleRange(wr, minTUnix, toPromModelTime(leaseStart))
+	case hasBackfill && !allowInsert:
+		filterOutSampleRange(wr, toPromModelTime(leaseStart), maxTUnix+1) //maxTUnix+1 because filterSamples treats end as exclusive.
+	case !hasBackfill && !allowInsert:
+		wr.Timeseries = wr.Timeseries[:0]
+		log.Debug("msg", "the samples aren't from the leader prom instance. skipping the insert")
+	default:
+		//nothing to do
+	}
+	return nil
+}
+
+func toPromModelTime(t time.Time) int64 {
+	return int64(model.TimeFromUnixNano(t.UnixNano()))
+}
+
+func (h *Filter) filterBackfill(wr *prompb.WriteRequest, minT, maxT time.Time, cluster, replica string) (bool, error) {
+	hasBackfill := false
+	backfillStart := minT
+	for {
+		if !backfillStart.Before(maxT) {
+			break
+		}
+
+		keepRangeStart, keepRangeEnd, err := h.service.GetBackfillLeaseRange(backfillStart, maxT, cluster, replica)
+		if err != nil {
+			return false, fmt.Errorf("could not check backfill ha lease: %#v", err)
+		}
+		// Nothing else to keep.
+		if keepRangeStart.IsZero() || keepRangeEnd.IsZero() {
+			break
 		}
+
+		hasBackfill = true
+		// Filter out all samples before the keep range.
+		if !keepRangeStart.Before(backfillStart) {
+			filterOutSampleRange(
+				wr,
+				toPromModelTime(backfillStart),
+				toPromModelTime(keepRangeStart),
+			)
+		}
+
+		backfillStart = keepRangeEnd
+	}
+
+	// If we had backfill, we need to filter out the range after the keep range
+	// till the start of the lease (end of possible backfill).
+	if hasBackfill {
+		filterOutSampleRange(
+			wr,
+			toPromModelTime(backfillStart),
+			toPromModelTime(maxT),
+		)
+	}
+
+	return hasBackfill, nil
+}
+
+func filterOutSampleRange(wr *prompb.WriteRequest, timeStart, timeEnd int64) {
+	for i := range wr.Timeseries {
+		t := &wr.Timeseries[i]
+		numAccepted := 0
+		for j := range t.Samples {
+			sample := t.Samples[j]
+			if sample.Timestamp >= timeStart && sample.Timestamp < timeEnd {
+				continue
+			}
+			t.Samples[numAccepted] = sample
+			numAccepted++
+		}
+		for j := numAccepted; j < len(t.Samples); j++ {
+			t.Samples[j] = prompb.Sample{}
+		}
+		t.Samples = t.Samples[:numAccepted]
+	}
+}
+
+func filterLabels(wr *prompb.WriteRequest) {
+	numAccepted := 0
+	for i := range wr.Timeseries {
+		t := &wr.Timeseries[i]
 		if len(t.Samples) == 0 {
-			// Empty the labels for posterity.
-			t.Labels = make([]prompb.Label, 0)
 			continue
 		}
-
+		wr.Timeseries[numAccepted] = *t
+		numAccepted++
 		// Drop __replica__ labelSet from samples,
 		// we don't want samples from the same Prometheus
 		// HA set to become different series.
 		for ind, value := range t.Labels {
-			if value.Name == ReplicaNameLabel && value.Value == replicaName {
+			if value.Name == ReplicaNameLabel {
 				t.Labels = append(t.Labels[:ind], t.Labels[ind+1:]...)
 				break
 			}
 		}
 	}
-
-	return nil
+	wr.Timeseries = wr.Timeseries[:numAccepted]
```

Comment:
don't we need to clear the memory of the tail of wr.Timeseries for memory purposes. The equivalent of line 146-148. 
like 
```
		for j := numAccepted; j < len(wr.Timeseries); j++ {
			wr.Timeseries[j] = prompb.Timeseries{}
		}
```

Reply:
This is done in the deferred function `filterLabels` which needs a rename since that's not all what it all does.

That function will called at the end of this one to remove HA replica labels and remove any `TimeSeries` which don't contain any samples.

Will changing the name to `finalFiltering` be good enough? Will add a comment what that actually means.

Reply:
I am confused isn't this inside finalFiltering? What I mean here is how do all the timeseries after numAccepted get cleared?

Reply:
You are right, I don't know what I was thinking. Added it to the filtering function.